src/content/reference/react-dom/hooks/useFormState.md

```diff
@@ -0,0 +1,219 @@
+      {message}
+    </form>
+  );
+}
+
+export default function App() {
+  return (
+    <>
+      <AddToCartForm itemID="1" itemTitle="Javascript: The Definitive Guide" />
+      <AddToCartForm itemID="2" itemTitle="Javascript: The Good Parts" />
+    </>
+  )
+}
+```
+
+```js actions.js
+"use server";
+
+export async function addToCart(prevState, queryData) {
+  const itemID = queryData.get('itemID');
+  if (itemID === "1") {
+    return "Added to cart";
+  } else {
+    return "Couldn't add to cart: the item is sold out.";
+  }
+}
+```
+
+```css styles.css hidden
+form {
+  border: solid 1px black;
+  margin-bottom: 24px;
+  padding: 12px
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "experimental",
+    "react-dom": "experimental",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+</Sandpack>  
+
+
+### Display structured information after submitting a form {/*display-structured-information-after-submitting-a-form*/}
+
+The return value from a server action can be any serializable value. For example, it could be an object that includes a boolean indicating whether the action was successful, an error message, or updated information.
+
+<Sandpack>
+
+```js App.js
+import { useState } from "react";
+import { experimental_useFormState as useFormState } from "react-dom";
+import { addToCart } from "./actions.js";
+
+function AddToCartForm({itemID, itemTitle}) {
+  const [formState, formAction] = useFormState(addToCart, {});
+  return (
+    <form action={formAction}>
+      <h2>{itemTitle}</h2>
+      <input type="hidden" name="itemID" value={itemID} />
+      <button type="submit">Add to Cart</button>
+      {formState?.success &&
+        <div className="toast">
+          Added to cart! Your cart now has {formState.cartSize} items.
+        </div>
+      }
+      {formState?.success === false &&
+        <div className="error">
+          Failed to add to cart: {formState.message}
+        </div>
+      }
+    </form>
+  );
+}
+
+export default function App() {
+  return (
+    <>
+      <AddToCartForm itemID="1" itemTitle="Javascript: The Definitive Guide" />
+      <AddToCartForm itemID="2" itemTitle="Javascript: The Good Parts" />
+    </>
+  )
+}
+```
+
+```js actions.js
+"use server";
+
+export async function addToCart(prevState, queryData) {
+  const itemID = queryData.get('itemID');
+  if (itemID === "1") {
+    return {
+      success: true,
+      cartSize: 12,
+    };
+  } else {
+    return {
+      success: false,
+      message: "The item is sold out.",
+    };
+  }
+}
+```
+
+```css styles.css hidden
+form {
+  border: solid 1px black;
+  margin-bottom: 24px;
+  padding: 12px
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "experimental",
+    "react-dom": "experimental",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+</Sandpack>  
```

Comment:
Have you considered adding an example demonstrating how you could use useFormState to update the UI without JavaScript?

Reply:
I don't know how to demonstrate that without having RSC on the docs website, do you have an idea of how that would look?

Reply:
You could add a non-interactive example with standard Markdown code syntax e.g. `` ```js ``

Reply:
Is there some specific application or use case for updating the UI without javascript that is different from existing examples such as Add to Cart?